bridges/snowbridge/pallets/system-frontend/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+//! Frontend which will be deployed on AssetHub for calling the V2 system pallet
+//! on BridgeHub.
+//!
+//! # Extrinsics
+//!
+//! * [`Call::create_agent`]: Create agent for any kind of sovereign location on Polkadot network.
+//! * [`Call::register_token`]: Register Polkadot native asset as a wrapped ERC20 token on Ethereum.
+#![cfg_attr(not(feature = "std"), no_std)]
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+pub use weights::*;
+
+use frame_support::{pallet_prelude::*, traits::EnsureOrigin};
+use frame_system::pallet_prelude::*;
+use sp_std::prelude::*;
+use xcm::prelude::*;
+use xcm_executor::traits::TransactAsset;
+
+#[cfg(feature = "runtime-benchmarks")]
+use frame_support::traits::OriginTrait;
+
+pub use pallet::*;
+use snowbridge_core::AssetMetadata;
+
+#[derive(Encode, Decode, Debug, PartialEq, Clone, TypeInfo)]
+pub enum ControlCall {
+	#[codec(index = 1)]
+	CreateAgent { location: Box<VersionedLocation>, fee: u128 },
+	#[codec(index = 2)]
+	RegisterToken { asset_id: Box<VersionedLocation>, metadata: AssetMetadata, fee: u128 },
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Encode, Decode, Debug, PartialEq, Clone, TypeInfo)]
+pub enum SnowbridgeControl {
+	#[codec(index = 90)]
+	Control(ControlCall),
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+pub trait BenchmarkHelper<O>
+where
+	O: OriginTrait,
+{
+	fn make_xcm_origin(location: Location) -> O;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Origin check for XCM locations that can create agents
+		type CreateAgentOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
+
+		/// Origin check for XCM locations that can create agents
+		type RegisterTokenOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
+
+		/// XCM message sender
+		type XcmSender: SendXcm;
+
+		/// To withdraw and deposit an asset.
+		type AssetTransactor: TransactAsset;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		type Helper: BenchmarkHelper<Self::RuntimeOrigin>;
+
+		type FeeAsset: Get<Location>;
+
+		type DeliveryFee: Get<Asset>;
+
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A `CreateAgent` message was sent to Bridge Hub
+		CreateAgent { location: Location },
+		/// A message to register a Polkadot-native token was sent to Bridge Hub
+		RegisterToken {
+			/// Location of Polkadot-native token
+			location: Location,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Convert versioned location failure
+		UnsupportedLocationVersion,
+		/// Check location failure, should start from the dispatch origin as owner
+		OwnerCheck,
```

Comment:
@yrong 

```suggestion
		InvalidAssetOwner,
```

Reply:
Also addressed in https://github.com/vgeddes/polkadot-sdk/pull/6/